# Overview:

* In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
* The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas.

In [1]:
import pandas as pd 
import numpy as np

In [2]:
df = pd.read_csv('autos.csv', encoding = 'Latin-1')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [3]:
df.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Cleaning columns name:

In [4]:
df.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
df.columns = (['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'])
df.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuelType  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Here I changed the name of the columns that were a bit confusing. I made the columns name easier to read and understand.

### Initial Exploration and Cleaning:

In [6]:
df.describe()

registration_year       powerPS  registration_month  nrOfPictures  \
count       50000.000000  50000.000000        50000.000000       50000.0   
mean         2005.073280    116.355920            5.723360           0.0   
std           105.712813    209.216627            3.711984           0.0   
min          1000.000000      0.000000            0.000000           0.0   
25%          1999.000000     70.000000            3.000000           0.0   
50%          2003.000000    105.000000            6.000000           0.0   
75%          2008.000000    150.000000            9.000000           0.0   
max          9999.000000  17700.000000           12.000000           0.0   

         postalCode  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [7]:
df.describe(include = 'O')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   

       vehicleType  gearbox  model   odometer fuelType       brand  \
count        44905    47320  47242      50000    45518       50000   
unique           8        2    245         13        7          40   
top      limousine  manuell   golf  150,000km   benzin  volkswagen   
freq         12859    36993   4024      32424    30107       10687   

       unrepaired_damage           ad_created             lastSeen  
count              40171                50000                50000  
unique                 2                   76                39481  
top                 nein  2016-04-03 00:00:00  2016-04-07 06:17:27  
freq               35232                 1946                    8

* Any columns that have mostly one value are candidates to be dropped 
* Columns like 'price' and 'odometer' ara numeric columns sotred as object, so we need to covert it:

In [8]:
df.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuelType  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [9]:
df['price'] = df['price'].str.replace('$', '').str.replace(',','')
df['price'] = df['price'].astype(float)

df['odometer'] = df['odometer'].str.replace('km', '').str.replace(',','')
df['odometer'] = df['odometer'].astype(float)

C:\Users\Henrique\AppData\Local\Temp\ipykernel_8056\957285289.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df['price'] = df['price'].str.replace('$', '').str.replace(',','')


In [10]:
df['price'].dtypes

dtype('float64')

* Changing column´s name for better viewing:

In [11]:
df.rename({'odometer': 'odometer_km', 'dateCrawled': 'date_crawled', 'price': 'price_usd', 'lastSeen': 'last_seen'}, axis = 1, inplace = True)

### Exploring the Odometer and Price Columns:

In [12]:
print(df.price_usd.unique().shape) #2357 unique values
#priting the info without cientific notation
with pd.option_context('display.float_format', '{:.2f}'.format):
    print(df.price_usd.describe())

(2357,)
count      50000.00
mean        9840.04
std       481104.38
min            0.00
25%         1100.00
50%         2950.00
75%         7200.00
max     99999999.00
Name: price_usd, dtype: float64


In [13]:
#getting the maximum prices 
df['price_usd'].value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price_usd, dtype: int64

In [14]:
#getting the minimum prices
df['price_usd'].value_counts().sort_index(ascending=True).head(20)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price_usd, dtype: int64

After checking the minimum prices and the max in 'price' column, it can be observed that are some strange values, unlikely prices. A car, in USD doesn´t cost less than 200 and i is very unlikely to cost more than 350k. Knowing that, I decided to use these measures on my non-outliers´ filter.

In [15]:
price_clean = df[df['price_usd'].between(250, 350000)]
print(len(price_clean)) #47288 cars in this range

47288


In [16]:
#checking odometer 
print(df.odometer_km.unique().shape) #13 unique values
#priting the info withou cientific notation
with pd.option_context('display.float_format', '{:.2f}'.format):
    print(df.odometer_km.describe())

(13,)
count    50000.00
mean    125732.70
std      40042.21
min       5000.00
25%     125000.00
50%     150000.00
75%     150000.00
max     150000.00
Name: odometer_km, dtype: float64


Values seem to represent reality

* Droping columns that have small unique values:

In [17]:
df = df.drop(['seller', 'offerType', 'abtest', 'gearbox'],axis = 1)

### Exploring the date columns:
* Here I will transform the first 10 characters (2016-03-12) into numerical using **Series.value_counts()** and sorting by index

In [18]:
#date_crawled
print(df['date_crawled'])
print(df['date_crawled'].str[:10].value_counts(normalize = True, dropna = False)
     .sort_index(ascending = True))

0        2016-03-26 17:47:46
1        2016-04-04 13:38:56
2        2016-03-26 18:57:24
3        2016-03-12 16:58:10
4        2016-04-01 14:38:50
                ...         
49995    2016-03-27 14:38:19
49996    2016-03-28 10:50:25
49997    2016-04-02 14:44:48
49998    2016-03-08 19:25:42
49999    2016-03-14 00:42:12
Name: date_crawled, Length: 50000, dtype: object
2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.

In [19]:
df.columns

Index(['date_crawled', 'name', 'price_usd', 'vehicleType', 'registration_year',
       'powerPS', 'model', 'odometer_km', 'registration_month', 'fuelType',
       'brand', 'unrepaired_damage', 'ad_created', 'nrOfPictures',
       'postalCode', 'last_seen'],
      dtype='object')

In [20]:
#ad_created
print(df['ad_created'])
print(df['ad_created'].str[:10].value_counts(normalize = True, dropna = False)
      .sort_index(ascending = True))

0        2016-03-26 00:00:00
1        2016-04-04 00:00:00
2        2016-03-26 00:00:00
3        2016-03-12 00:00:00
4        2016-04-01 00:00:00
                ...         
49995    2016-03-27 00:00:00
49996    2016-03-28 00:00:00
49997    2016-04-02 00:00:00
49998    2016-03-08 00:00:00
49999    2016-03-13 00:00:00
Name: ad_created, Length: 50000, dtype: object
2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64


In [21]:
#last_seen 
print(df['last_seen'])
print(df['last_seen'].str[:10].value_counts(normalize = True, dropna = False)
     .sort_index(ascending = True))

0        2016-04-06 06:45:54
1        2016-04-06 14:45:08
2        2016-04-06 20:15:37
3        2016-03-15 03:16:28
4        2016-04-01 14:38:50
                ...         
49995    2016-04-01 13:47:40
49996    2016-04-02 14:18:02
49997    2016-04-04 11:47:27
49998    2016-04-05 16:45:07
49999    2016-04-06 21:18:48
Name: last_seen, Length: 50000, dtype: object
2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.024

* checking distribuition of 'registation_year'

In [22]:
df['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Aparently there is some error in the dataset, as we can see the maximum year is 9999, and that is impossible since we are in 2023. Also the minimum year is 1000, which is impossible since there were no cars in the year 1000.

In [23]:
df['registration_year'].unique()

array([2004, 1997, 2009, 2007, 2003, 2006, 1995, 1998, 2000, 2017, 2010,
       1999, 1982, 1990, 2015, 2014, 1996, 1992, 2005, 2002, 2012, 2011,
       2008, 1985, 2016, 1994, 1986, 2001, 2018, 2013, 1972, 1993, 1988,
       1989, 1967, 1973, 1956, 1976, 4500, 1987, 1991, 1983, 1960, 1969,
       1950, 1978, 1980, 1984, 1963, 1977, 1961, 1968, 1934, 1965, 1971,
       1966, 1979, 1981, 1970, 1974, 1910, 1975, 5000, 4100, 2019, 1959,
       9996, 9999, 6200, 1964, 1958, 1800, 1948, 1931, 1943, 9000, 1941,
       1962, 1927, 1937, 1929, 1000, 1957, 1952, 1111, 1955, 1939, 8888,
       1954, 1938, 2800, 5911, 1500, 1953, 1951, 4800, 1001], dtype=int64)

* By analizing the data, the most acceptable year for maximum is 2019 and minimum is 1927

* I´ll change the data, by doing the following: cars that have year below 1927 will receive 1927 and years above 2019 will get 2019

In [24]:
print(df['registration_year'].value_counts(normalize = True).head(50).sort_index(ascending = True))

1960    0.00068
1967    0.00054
1968    0.00052
1970    0.00090
1971    0.00054
1972    0.00070
1973    0.00052
1974    0.00048
1976    0.00054
1978    0.00094
1979    0.00070
1980    0.00194
1981    0.00062
1982    0.00086
1983    0.00106
1984    0.00106
1985    0.00210
1986    0.00152
1987    0.00150
1988    0.00284
1989    0.00362
1990    0.00790
1991    0.00712
1992    0.00782
1993    0.00890
1994    0.01320
1995    0.02626
1996    0.02888
1997    0.04056
1998    0.04906
1999    0.06000
2000    0.06708
2001    0.05406
2002    0.05066
2003    0.05454
2004    0.05474
2005    0.06030
2006    0.05416
2007    0.04608
2008    0.04462
2009    0.04196
2010    0.03194
2011    0.03268
2012    0.02646
2013    0.01612
2014    0.01332
2015    0.00798
2016    0.02632
2017    0.02906
2018    0.00984
Name: registration_year, dtype: float64


In [25]:
#creating a function to get correct values 

def modify_registration_year(value):
    if value < 1927:
        return 1927
    elif value > 2019:
        return 2019
    else:
        return value 

#apllying the function created above 
df['registration_year'] = df['registration_year'].apply(modify_registration_year)

df['registration_year'].unique() 

array([2004, 1997, 2009, 2007, 2003, 2006, 1995, 1998, 2000, 2017, 2010,
       1999, 1982, 1990, 2015, 2014, 1996, 1992, 2005, 2002, 2012, 2011,
       2008, 1985, 2016, 1994, 1986, 2001, 2018, 2013, 1972, 1993, 1988,
       1989, 1967, 1973, 1956, 1976, 2019, 1987, 1991, 1983, 1960, 1969,
       1950, 1978, 1980, 1984, 1963, 1977, 1961, 1968, 1934, 1965, 1971,
       1966, 1979, 1981, 1970, 1974, 1927, 1975, 1959, 1964, 1958, 1948,
       1931, 1943, 1941, 1962, 1937, 1929, 1957, 1952, 1955, 1939, 1954,
       1938, 1953, 1951], dtype=int64)

In [26]:
#checking if there is any value below 1927
count1 = df['registration_year'] < 1927
count1.value_counts()

False    50000
Name: registration_year, dtype: int64

In [27]:
count2 = df['registration_year'] > 2019
count2.value_counts()

False    50000
Name: registration_year, dtype: int64

* Now that the data was cleaned, we don´t have values under the year of 1927 and above 2019

### Exploring Price by Brand:

In [28]:
df.brand.unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [29]:
#Cheking out the brand with most cars in the market 
df.brand.value_counts()

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
smart               701
citroen             701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
honda               399
kia                 356
alfa_romeo          329
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                110
subaru              109
land_rover           99
saab                 80
daewoo               79
trabant              78
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

In [30]:
#getting it in percentage
df.brand.value_counts(normalize = True)*100

volkswagen        21.374
opel              10.922
bmw               10.858
mercedes_benz      9.468
audi               8.566
ford               6.958
renault            4.808
peugeot            2.912
fiat               2.616
seat               1.882
skoda              1.572
mazda              1.514
nissan             1.508
smart              1.402
citroen            1.402
toyota             1.234
sonstige_autos     1.092
hyundai            0.976
volvo              0.914
mini               0.848
mitsubishi         0.812
honda              0.798
kia                0.712
alfa_romeo         0.658
porsche            0.588
suzuki             0.586
chevrolet          0.566
chrysler           0.362
dacia              0.258
daihatsu           0.256
jeep               0.220
subaru             0.218
land_rover         0.198
saab               0.160
daewoo             0.158
trabant            0.156
jaguar             0.154
rover              0.138
lancia             0.114
lada               0.062


* Volswagen is the most popular car on Ebay leading with 21% of the used cars.
* Now that we have the percentages, we can analyze the pricec of cars that have more than 5% of the market share.

In [31]:
more_than_5 = (df['brand'].value_counts(normalize = True) * 100) >5
print(more_than_5.head(6))

volkswagen       True
opel             True
bmw              True
mercedes_benz    True
audi             True
ford             True
Name: brand, dtype: bool


In [32]:
mean_by_brand = {}

unique_brands = df['brand'].unique()

for i in unique_brands:
    selected_row = df[df['brand'] == i]
    if more_than_5[i] == True:
        mean = selected_row['price_usd'].mean()
        mean_by_brand[i] = mean
        
dict(sorted(mean_by_brand.items(), key=lambda x: x[1], reverse=True))

{'mercedes_benz': 29511.955428812842,
 'audi': 8965.560354891431,
 'bmw': 8252.918953766808,
 'ford': 7105.662546708824,
 'volkswagen': 6384.167399644428,
 'opel': 5106.092657022524}

As you can see, among the top 6 brands in ebay, **Mercedez Benz** is the most expensive car´s brand on ebay with almost USD 30,000 as average, followed by Audi with an average of almost USD 9,000. On the other hand, **Opel** is the cheapest costing around USD 5,106.

### Storing Aggregate Data in DataFrame:
* If we are to buy a car on eBay platform, I thinnk we could try to find the cars that still in good shape. In order to know which Brand sell the cars that are in good shape we have to consider the mileage. Let us not take a look on the mileage and also try to find if the pricing of cars depend on the mileage.

In [34]:
mileage_by_brand={}
unique_brand=df['brand'].unique()

for l in unique_brand:
    sel_brand=df[df['brand'] == l]
    if more_than_5[l]==True:
        mileage_by_brand[l]=sel_brand['odometer_km'].mean()
        
dict(sorted(mileage_by_brand.items(), key=lambda x: x[1], reverse=True))

{'bmw': 132521.64302818198,
 'mercedes_benz': 130886.14279678918,
 'audi': 129643.9411627364,
 'opel': 129298.66324848929,
 'volkswagen': 128955.27276129878,
 'ford': 124131.93446392642}

In [37]:
#Converting the dictionary into dataframe series

mbp_series = pd.Series(mean_by_brand)
pd.DataFrame(mbp_series, columns = ['mean_price'])

mbo_series = pd.Series(mileage_by_brand)
pd.DataFrame(mbo_series, columns = ['mean_milage_brand'])

mean_price = pd.Series(mean_by_brand).sort_values(ascending = False)
mean_milage_brand = pd.Series(mileage_by_brand).sort_values(ascending = False)


print('Price by Brand')
print('\n')
print(mbp_series)
print('\n')
print('Mileage By Brand')
print('\n')
print(mbo_series)

Price by Brand


bmw               8252.918954
volkswagen        6384.167400
ford              7105.662547
mercedes_benz    29511.955429
audi              8965.560355
opel              5106.092657
dtype: float64


Mileage By Brand


bmw              132521.643028
volkswagen       128955.272761
ford             124131.934464
mercedes_benz    130886.142797
audi             129643.941163
opel             129298.663248
dtype: float64


In [40]:
brand_info = pd.DataFrame(mbo_series, columns = ['mean_mileage_brand'])
brand_info

mean_mileage_brand
bmw                 132521.643028
volkswagen          128955.272761
ford                124131.934464
mercedes_benz       130886.142797
audi                129643.941163
opel                129298.663248

In [44]:
brand_info['mean_price'] = mbp_series
brand_info

mean_mileage_brand    mean_price
bmw                 132521.643028   8252.918954
volkswagen          128955.272761   6384.167400
ford                124131.934464   7105.662547
mercedes_benz       130886.142797  29511.955429
audi                129643.941163   8965.560355
opel                129298.663248   5106.092657

# Conclusion:
In this project, we worked with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. We started by cleaning our data then we evaluated which brands are the most expensive. We come to realized that among the six most popular brands on eBay platform; Mercedes Benz, Audi, and BMW are the most expensive brand. Volkswagen has a middle price range, while Ford and Opel are the cheapest.

If we are to buy a car on eBay platform we should not only consider price but also the car that still in good shape. So we went on and evaluate if the pricing of cars on eBay platform is being influenced by the mileage. Our result prove that the mileage has no influence on pricing the cars on eBay platform. As seen above, Opel brand has lowest average mileage compared to other brand but people sell them at the low price while Mercedes benz has the second high average mileage but still the most expensive brand on eBay platform.